crates/wasm-smith/src/component.rs

```diff
@@ -0,0 +1,802 @@
+        let mut variants = vec![];
+        let mut variant_names = HashSet::new();
+        arbitrary_loop(u, 0, 100, |u| {
+            variants.push(crate::unique_string(100, &mut variant_names, u)?);
+            Ok(true)
+        })?;
+        Ok(UnionType { variants })
+    }
+
+    fn arbitrary_optional_type(&mut self, u: &mut Unstructured<'a>) -> Result<OptionalType> {
+        Ok(OptionalType {
+            inner_ty: self.arbitrary_value_type(u)?,
+        })
+    }
+
+    fn arbitrary_expected_type(&mut self, u: &mut Unstructured<'a>) -> Result<ExpectedType> {
+        Ok(ExpectedType {
+            ok_ty: self.arbitrary_value_type(u)?,
+            err_ty: self.arbitrary_value_type(u)?,
+        })
+    }
+
+    fn arbitrary_compound_type(&mut self, u: &mut Unstructured<'a>) -> Result<CompoundType> {
+        match u.int_in_range(0..=9)? {
+            0 => Ok(CompoundType::Record(self.arbitrary_record_type(u)?)),
+            1 => Ok(CompoundType::Variant(self.arbitrary_variant_type(u)?)),
+            2 => Ok(CompoundType::List(self.arbitrary_list_type(u)?)),
+            3 => Ok(CompoundType::Tuple(self.arbitrary_tuple_type(u)?)),
+            4 => Ok(CompoundType::Flags(self.arbitrary_flags_type(u)?)),
+            5 => Ok(CompoundType::Enum(self.arbitrary_enum_type(u)?)),
+            6 => Ok(CompoundType::Union(self.arbitrary_union_type(u)?)),
+            7 => Ok(CompoundType::Optional(self.arbitrary_optional_type(u)?)),
+            8 => Ok(CompoundType::Expected(self.arbitrary_expected_type(u)?)),
+            9 => Ok(CompoundType::Named(
+                self.arbitrary_named_type(u, &mut HashSet::new())?,
+            )),
+            _ => unreachable!(),
+        }
+    }
+
+    fn arbitrary_import_section(&mut self, u: &mut Unstructured<'a>) -> Result<()> {
+        todo!()
+    }
+
+    fn arbitrary_func_section(&mut self, u: &mut Unstructured<'a>) -> Result<()> {
+        todo!()
+    }
+
+    fn arbitrary_core_section(&mut self, u: &mut Unstructured<'a>) -> Result<()> {
+        todo!()
+    }
+
+    fn arbitrary_component_section(&mut self, u: &mut Unstructured<'a>) -> Result<()> {
+        todo!()
+    }
+
+    fn arbitrary_instance_section(&mut self, u: &mut Unstructured<'a>) -> Result<()> {
+        todo!()
+    }
+
+    fn arbitrary_export_section(&mut self, u: &mut Unstructured<'a>) -> Result<()> {
+        todo!()
+    }
+
+    fn arbitrary_start_section(&mut self, u: &mut Unstructured<'a>) -> Result<()> {
+        todo!()
+    }
+
+    fn arbitrary_alias_section(&mut self, u: &mut Unstructured<'a>) -> Result<()> {
+        todo!()
+    }
+}
+
+#[derive(Debug)]
+enum Section<'a> {
+    Custom(CustomSection<'a>),
+    Type(TypeSection),
+    Import(ImportSection),
+    Func(FuncSection),
+    Core(CoreSection),
+    Component(ComponentSection),
+    Instance(InstanceSection),
+    Export(ExportSection),
+    Start(StartSection),
+    Alias(AliasSection),
+}
+
+#[derive(Debug)]
+struct CustomSection<'a> {
```

Comment:
No strong feelings either way. I haven't found the `'a` to be annoying yet, and figured we might as well avoid a copy if it is easy. But I also don't feel strongly enough to push back on copying.

Reply:
I'm mostly worried about consumers like wasmtime where threading `'a` around when it's not already present can be pretty onerous and otherwise it doesn't seem all that important for perf here.

Reply:
Makes sense, I'll remove the `'a`.